python/taichi/lang/transformer.py

```diff
@@ -32,18 +32,27 @@ def __init__(self,
                  arg_features=None):
         super().__init__()
         self.local_scopes = []
+        self.control_scopes = []
         self.excluded_parameters = excluded_paremeters
         self.is_kernel = is_kernel
         self.func = func
         self.arg_features = arg_features
         self.returns = None
 
+    # e.g.: FunctionDef, Module, Global
     def variable_scope(self, *args):
-        return ScopeGuard(self, *args)
+        return ScopeGuard(self.local_scopes, *args)
+
+    # e.g.: For, While
+    def control_scope(self, *args):
```

Comment:
Nit: `ScopeGuard` is specialized for local variables, because its `__exit__` has this operation to delete the local variables in the scope, which doesn't make sense to control scope. I think the easiest thing to prevent errors is just to remove `*args` here, so that `stmt_block` will be `None` for control scopes.